x/auth/types/account_retriever.go

```diff
@@ -1,67 +1,65 @@
 package types
 
 import (
-	"fmt"
+	context "context"
 
 	"github.com/cosmos/cosmos-sdk/client"
-	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/cosmos/cosmos-sdk/codec/types"
 	sdk "github.com/cosmos/cosmos-sdk/types"
 )
 
 // AccountRetriever defines the properties of a type that can be used to
 // retrieve accounts.
 type AccountRetriever struct {
-	codec codec.JSONMarshaler
+	ir types.InterfaceRegistry
 }
 
-// NewAccountRetriever initialises a new AccountRetriever instance.
-func NewAccountRetriever(codec codec.JSONMarshaler) AccountRetriever {
-	return AccountRetriever{codec: codec}
+// NewAccountRetriever initializes a new AccountRetriever instance.
+func NewAccountRetriever(ir types.InterfaceRegistry) AccountRetriever {
+	return AccountRetriever{ir: ir}
 }
 
 // GetAccount queries for an account given an address and a block height. An
 // error is returned if the query or decoding fails.
-func (ar AccountRetriever) GetAccount(querier client.NodeQuerier, addr sdk.AccAddress) (AccountI, error) {
-	account, _, err := ar.GetAccountWithHeight(querier, addr)
+func (ar AccountRetriever) GetAccount(clientCtx client.Context, addr sdk.AccAddress) (AccountI, error) {
+	account, _, err := ar.GetAccountWithHeight(clientCtx, addr)
 	return account, err
 }
 
 // GetAccountWithHeight queries for an account given an address. Returns the
 // height of the query with the account. An error is returned if the query
 // or decoding fails.
-func (ar AccountRetriever) GetAccountWithHeight(querier client.NodeQuerier, addr sdk.AccAddress) (AccountI, int64, error) {
-	bs, err := ar.codec.MarshalJSON(QueryAccountRequest{Address: addr})
+func (ar AccountRetriever) GetAccountWithHeight(clientCtx client.Context, addr sdk.AccAddress) (AccountI, int64, error) {
+	queryClient := NewQueryClient(clientCtx)
+	res, err := queryClient.Account(context.Background(), &QueryAccountRequest{Address: addr})
 	if err != nil {
 		return nil, 0, err
 	}
 
-	bz, height, err := querier.QueryWithData(fmt.Sprintf("custom/%s/%s", QuerierRoute, QueryAccount), bs)
-	if err != nil {
-		return nil, height, err
-	}
-
-	var account AccountI
-	if err := ar.codec.UnmarshalJSON(bz, &account); err != nil {
-		return nil, height, err
+	var acc AccountI
+	if err := ar.ir.UnpackAny(res.Account, &acc); err != nil {
+		return nil, 0, err
 	}
 
-	return account, height, nil
+	return acc, 0, nil
```

Comment:
@aaronc, two questions:

1. Does a gRPC query appropriately handle a custom height set on `client.Context`?
2. How do we return the height from a successful gRPC query?

Reply:
See `GRPCBlockHeightHeader`. That does both of those things as I understand it. @amaurymartiny added it and can maybe say more.

Reply:
Correct, is you pass this header, the gRPC server will parse it, and create a sdk.Context at that height.

https://github.com/cosmos/cosmos-sdk/blob/e9534b093575f4d7edb201ba9b2c3738c7e4ad14/baseapp/grpcserver.go#L31-L42

Reply:
And for 2/, we return the height as a header metadata too:

https://github.com/cosmos/cosmos-sdk/blob/e9534b093575f4d7edb201ba9b2c3738c7e4ad14/baseapp/grpcserver.go#L54

Reply:
Thanks @amaurymartiny, but how do I actually get the height? All I have access to is the `QueryAccountResponse` and `QueryClient`.

Reply:
@alexanderbez I think this could help: https://github.com/grpc/grpc-go/blob/master/Documentation/grpc-metadata.md#receiving-metadata

Here's also a test I wrote: https://github.com/cosmos/cosmos-sdk/blob/master/server/grpc/server_test.go#L73-L80

Reply:
Ahhh yes, this is helpful, thank you.

Just want to clarify, this behavior is the same as it was before correct? Meaning:

1. If I specify a height on the `clientCtx`, I'll get a response with that height.
2. If I don't specify a height on the `clientCtx`, I'll get a response with the current/latest height.

Reply:
Correct 👍.

Reply:
`clientCtx` is how clients provide an explicit height.

e.g.
```go
clientCtx := client.GetClientContextFromCmd(cmd)
clientCtx, err := client.ReadQueryCommandFlags(clientCtx, cmd.Flags())
if err != nil {
    return err
}
```

Here `clientCtx.Height` is set if the user did a `--height=H`. The previous behavior was that when the query was handled, the `clientCtx.Height` would be provided to the `Query` object that `BaseApp#Query` processes (using the latest height when omitted).

Is this no longer the case?! Heights have to explicitly set on the `queryClient.*` call???

Reply:
Just took a look at BaseApp, looks like this behavior still holds:

```go
func (app *BaseApp) handleQueryGRPC(handler GRPCQueryHandler, req abci.RequestQuery) abci.ResponseQuery {
	ctx, err := app.createQueryContext(req.Height, req.Prove)
	if err != nil {
		return sdkerrors.QueryResult(err)
	}

	res, err := handler(ctx, req)
	if err != nil {
		res = sdkerrors.QueryResult(err)
		res.Height = req.Height
		return res
	}

	return res
}
```

Reply:
Ok, I read through the flow, and I think this still holds...ignore me or correct me if I'm wrong.

Reply:
Ah alright, yeah I got mixed up, what I was describing above was using the gRPC server.

In your case, since you use `clientCtx`, your gRPC request bypasses the gRPC server completely, and goes through tendermint query, so `clientCtx.Height` is indeed respected.

But yeah, the height is not returned in this case, see the [`_` on this line](https://github.com/cosmos/cosmos-sdk/blob/2f44fbf2ababbe1637341fef7ec367e829182dde/client/grpc_query.go#L25). Maybe in this `client.Context#Invoke` function, we could return the height in the header too. I can make a PR for that tomorrow.

Reply:
Yeah I'm getting a bit lost in this. We need to ensure height queries still work. Specifically, the two points I laid out in https://github.com/cosmos/cosmos-sdk/pull/7006#discussion_r468685317. This respects the current CLI framework that `--height` should work out-of-the-box.